baselines/fednova/fednova/strategy.py

```diff
@@ -0,0 +1,146 @@
+"""FedNova strategy."""
+
+from logging import INFO
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+from flwr.common import (
+    Metrics,
+    NDArray,
+    NDArrays,
+    Parameters,
+    Scalar,
+    ndarrays_to_parameters,
+    parameters_to_ndarrays,
+)
+from flwr.common.logger import log
+from flwr.common.typing import FitRes
+from flwr.server.client_proxy import ClientProxy
+from flwr.server.strategy import FedAvg
+from flwr.server.strategy.aggregate import aggregate
+from omegaconf import DictConfig
+
+
+class FedNova(FedAvg):
+    """FedNova."""
+
+    def __init__(self, exp_config: DictConfig, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.global_momentum_buffer: List[NDArray] = []
+        if self.initial_parameters is not None:
+            self.global_parameters: List[NDArray] = parameters_to_ndarrays(
+                self.initial_parameters
+            )
+
+        self.exp_config = exp_config
+        self.lr = exp_config.optimizer.lr
+        self.gmf = exp_config.optimizer.gmf
+        self.best_test_acc = 0.0
+
+    def aggregate_fit(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, FitRes]],
+        failures: List[Union[Tuple[ClientProxy, FitRes], BaseException]],
+    ):
+        """Aggregate the results from the clients."""
+        if not results:
+            return None, {}
+
+        # Do not aggregate if there are failures and failures are not accepted
+        if not self.accept_failures and failures:
+            return None, {}
+
+        local_tau = [res.metrics["tau"] for _, res in results]
+        tau_eff = np.sum(local_tau)
+
+        aggregate_parameters = []
+
+        for _client, res in results:
+            params = parameters_to_ndarrays(res.parameters)
+            scale = tau_eff / float(res.metrics["local_norm"])
+            scale *= float(res.metrics["weight"])
+            aggregate_parameters.append((params, scale))
+
+        agg_cum_gradient = aggregate(aggregate_parameters)
+        self.update_server_params(agg_cum_gradient)
+
+        return ndarrays_to_parameters(self.global_parameters), {}
+
+    def update_server_params(self, cum_grad: NDArrays):
```

Comment:
Could you add a couple of comments in this method to point to specific lines in the FedNova algorithm? (just to help people navigate the code more easily)

Reply:
Done. Added comments.